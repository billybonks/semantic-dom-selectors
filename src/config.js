class Config {
  constructor() {
    this.reset();
  }

  reset() {
    this.registeredFinders = [];
    this.defaultFinders = [];
    this.errorLevelOptions = {};
    this.customActors = {
      select: [],
      text: [],
      toggle: [],
      button: [],
    };
    this._rootElement = document.documentElement;
  }

  trim(text) {
    return text.replace(/\n/g, ' ').replace(/\s+/g, ' ').trim();
  }


  get rootElement() {
    return this._rootElement;
  }

  set rootElement(element) {
    if (element instanceof Element) {
      this._rootElement = element;
    } else {
      throw new Error(`rootElement Error: You tried to set root element to ${element}`);
    }
  }

  registerFinder(finder) {
    if (!this.errorLevelOptions[finder.key]) {
      this.errorLevelOptions[finder.key] = 1;
    }
    this.registeredFinders.push(finder);
  }

  registerActor({ type, run }) {
    if (!type) {
      throw new Error('You must specify the type of attribute of actor object');
    }
    if (!Object.keys(this.customActors).includes(type)) {
      throw new Error(`Unkown actor type ${type}`);
    }
    this.customActors[type].push(run);
  }

  setErrorLevels(config) {
    this.errorLevelOptions = config;
  }

  get actors() {
    return this.customActors;
  }

  get finders() {
    return this.registeredFinders.concat(this.defaultFinders);
  }

  get rules() {
    const rules = this.finders.reduce((acc, finder) => {
      const config = this.errorLevelOptions[finder.key];
      if (isNaN(config)) {
        acc[finder.key] = 1;
      } else {
        acc[finder.key] = config;
      }
      return acc;
    }, {});
    return rules;
  }
}

export default new Config();
