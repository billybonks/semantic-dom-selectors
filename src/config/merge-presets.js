import defaultPreset from '../presets/default';
import ariaPreset from '../presets/aria-compliant';

function uniq(array) {
  return Array.from(new Set(array));
}

const presets = {
  default: defaultPreset,
  'aria-compliant': ariaPreset,
};

export default function merge(config) {
  const empty = {
    finders: [],
    actors: [],
    rules: {},
  };

  let preset;
  if (config.preset) {
    preset = presets[config.preset];
    if (!preset) {
      throw new Error('unkown preset set for semantic-dom-selectors configuration');
    }
  } else {
    preset = {};
  }
  const normalizedPreset = Object.assign({}, empty, preset);
  const normalizedConfig = Object.assign({}, empty, config);
  const finders = uniq(normalizedPreset.finders.concat(normalizedConfig.finders));
  const actors = uniq(normalizedPreset.actors.concat(normalizedConfig.actors));
  const rules = Object.assign({}, normalizedPreset.rules, normalizedConfig.rules);
  return {
    finders,
    actors,
    rules,
  };
}
