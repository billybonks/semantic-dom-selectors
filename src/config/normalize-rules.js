const RULE_SEVERITY_STRINGS = ['error', 'warn', 'off'];
const RULE_SEVERITY = RULE_SEVERITY_STRINGS.reduce((map, value, index) => {
  map[value] = index;
  return map;
}, {});
const VALID_SEVERITIES = [0, 1, 2, 'off', 'warn', 'error'];

export default function normalizeRules(rules) {
  return Object.keys(rules).reduce((acc, ruleKey) => {
    let normalizedSeverity;
    const severity = rules[ruleKey];
    if (VALID_SEVERITIES.includes(severity)) {
      if (RULE_SEVERITY_STRINGS.includes(severity)) {
        normalizedSeverity = [RULE_SEVERITY[severity]];
      } else {
        normalizedSeverity = severity;
      }
    } else {
      throw new Error(`Invalid severity ${severity} set for ${ruleKey} please use ${VALID_SEVERITIES}`);
    }
    acc[ruleKey] = normalizedSeverity;
    return acc;
  }, {});
}
