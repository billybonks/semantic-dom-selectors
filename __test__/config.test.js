import config from '../src/config';

function render(html) {
  const element = document.createElement('div');
  const body = document.querySelector('body');
  element.innerHTML = html;
  body.appendChild(element);
}

function whenRootElementExists() {
  describe('when _rootElement exists', () => {
    let root = null;
    beforeEach(() => {
      root = document.createElement('div');
      config.rootElement = root;
    });

    afterEach(() => {
      config.rootElement = null;
    });
    it('returns rootElement', () => {
      expect(config.rootElement).toEqual(root);
    });
  });
}

describe('Configuration Module', () => {
  afterEach(() => {
    config.reset();
  });

  describe('rootElement configuration', () => {
    describe('setting rootElementSelector', () => {
      test('only allows strings', () => {
        expect(() => {
          config.rootElementSelector = 1;
        }).toThrowErrorMatchingSnapshot();
      });
      test('only allows valid selector strings', () => {
        expect(() => {
          config.rootElementSelector = 'asdasd::';
        }).toThrowErrorMatchingSnapshot();
      });
      test('only allows valid selector strings', () => {
        config.rootElementSelector = 'asdasd';
        expect(config.rootElementSelector).toEqual('asdasd');
      });
    });
    describe('setRootElement', () => {
      test('supports children of element', () => {
        const element = document.createElement('div');
        config.rootElement = element;
        expect(config._rootElement).toEqual(element);
      });
      test('does not support other trees extending node', () => {
        const element = document.createTextNode('asd');
        expect(() => {
          config.rootElement = element;
        }).toThrowErrorMatchingSnapshot();
      });
      test('does not support other random inputs', () => {
        expect(() => {
          config.rootElement = 'qqqq';
        }).toThrowErrorMatchingSnapshot();
      });
    });

    describe('Getting rootElement', () => {
      describe('When rootElementSelector is present', () => {
        beforeEach(() => {
          config.rootElementSelector = '#main-wrapper';
        });
        afterEach(() => {
          const body = document.querySelector('body');
          if (body.firstChild) {
            body.removeChild(body.firstChild);
          }
        });
        describe('Element exists on page', () => {
          beforeEach(() => {
            render('<div id=\'main-wrapper\'>');
          });
          whenRootElementExists();
          describe('when root element does not exist', () => {
            test('returns selector element', () => {
              expect(config.rootElement).toEqual(document.querySelector('#main-wrapper'));
            });
          });
        });
        describe('Element does not exist on page', () => {
          whenRootElementExists();
          test('returns document.documentElement', () => {
            expect(config.rootElement).toEqual(document.documentElement);
          });
        });
      });
    });
  });
  describe('Trim', () => {
    afterEach(() => {
      document.body.innerHTML = '';
    });

    test('it trims markup to single-spaced string', () => {
      const markup = `
        <button id="expected">
          <h3>Button Title</h3>
          <p>Description</p>
        </button>
      `;
      document.body.innerHTML = markup;
      expect(config.trim(document.body.innerHTML))
        .toEqual('<button id="expected"> <h3>Button Title</h3> <p>Description</p> </button>');
    });

    test('check dom', () => {
      expect(document.body.innerHTML).toEqual('');
    });
  });

  describe('Registering a finder', () => {
    it('adds it to registerdFinderArray', () => {
      config.registerFinder({
        key: 'custom-finder',
      });
      expect(config.registeredFinders.length).toEqual(1);
    });

    describe('sets a rule', () => {
      describe('no rule previously set', () => {
        it('adds a default rule with level of 1', () => {
          config.registerFinder({
            key: 'custom-finder',
          });
          expect(config.rules['custom-finder']).toEqual(1);
        });
      });

      describe('rule previously set', () => {
        it('does not override setting', () => {
          config.errorLevelOptions = {
            'custom-finder': 2,
          };
          config.registerFinder({
            key: 'custom-finder',
          });
          expect(config.rules['custom-finder']).toEqual(2);
        });
      });
    });
  });

  describe('Finder registration', () => {
    describe('Registering a finder', () => {
      it('adds it to registerdFinderArray', () => {
        config.registerFinder({
          key: 'custom-finder',
        });
        expect(config.registeredFinders.length).toEqual(1);
      });

      describe('sets a rule', () => {
        describe('no rule previously set', () => {
          it('adds a default rule with level of 1', () => {
            config.registerFinder({
              key: 'custom-finder',
            });
            expect(config.rules['custom-finder']).toEqual(1);
          });
        });

        describe('rule previously set', () => {
          it('does not override setting', () => {
            config.errorLevelOptions = {
              'custom-finder': 2,
            };
            config.registerFinder({
              key: 'custom-finder',
            });
            expect(config.rules['custom-finder']).toEqual(2);
          });
        });
      });
    });
  });

  describe('Actor registraition', () => {
    describe('if type is specified', () => {
      it('adds it to actor array', () => {
        config.registerActor({
          type: 'select',
          run: () => { },
        });
        expect(config.customActors.select.length).toEqual(1);
      });

      test('#getActors returns actors hash', () => {
        config.registerActor({
          type: 'select',
          run: () => { },
        });
        expect(config.actors).toEqual(config.customActors);
      });
    });

    describe('if type is not specified', () => {
      it('adds it to actor array', () => {
        expect(() => {
          config.registerActor({
            run: () => { },
          });
        }).toThrowErrorMatchingSnapshot();
      });
    });


    describe('if unkown type is specified', () => {
      it('adds it to actor array', () => {
        expect(() => {
          config.registerActor({
            type: 'nonsense',
            run: () => { },
          });
        }).toThrowErrorMatchingSnapshot();
      });
    });
  });

  describe('Rule customisation', () => {
    describe('--deprecations-- when using a deprecated rule name', () => {
      let rules;

      beforeEach(() => {
        rules = {
          ariaNotFound: 2,
          invalidFor: 1,
          perceivedByName: 0,
        };
      });

      it('notifies user that rule name is deprecated', () => {
        const originalWarn = console.warn;
        console.warn = jest.fn();
        config.remapDeprecatedRules(rules);
        expect(console.warn).toHaveBeenNthCalledWith(1, 'DEPRECATION: rule ariaNotFound name is deprecated please use not-aria-compliant');
        expect(console.warn).toHaveBeenNthCalledWith(2, 'DEPRECATION: rule invalidFor name is deprecated please use invalid-label-for');
        expect(console.warn).toHaveBeenNthCalledWith(3, 'DEPRECATION: rule perceivedByName name is deprecated please use name-attribute');
        console.warn = originalWarn;
      });

      it('maps rule to new name', () => {
        const result = config.remapDeprecatedRules(rules);
        expect(result).toEqual({
          'not-aria-compliant': 2,
          'invalid-label-for': 1,
          'name-attribute': 0,
        });
      });

      it('calls remapDeprecatedRules', () => {
        config.setErrorLevels({
          ariaNotFound: 2,
        });
        expect(config.errorLevelOptions).toEqual({
          'not-aria-compliant': 2,
        });
      });
    });

    test('you can whole sale set error levels', () => {
      config.setErrorLevels({
        'custom-finder': 2,
      });
      expect(config.errorLevelOptions).toEqual({
        'custom-finder': 2,
      });
    });
  });
});
