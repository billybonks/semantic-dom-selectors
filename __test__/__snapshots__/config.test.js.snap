// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Configuration Module Actor registraition if type is not specified adds it to actor array 1`] = `"You must specify the type of attribute of actor object"`;

exports[`Configuration Module Actor registraition if unkown type is specified adds it to actor array 1`] = `"Unkown actor type nonsense"`;

exports[`Configuration Module rootElement configuration setRootElement does not support other random inputs 1`] = `"rootElement Error: You tried to set root element to qqqq"`;

exports[`Configuration Module rootElement configuration setRootElement does not support other trees extending node 1`] = `"rootElement Error: You tried to set root element to [object Text]"`;

exports[`Configuration Module rootElement configuration setting rootElementSelector only allows strings 1`] = `"You Tried to set rootElementSelector to 1, but it was not a string"`;

exports[`Configuration Module rootElement configuration setting rootElementSelector only allows valid selector strings 1`] = `"You Tried to set rootElementSelector to asdasd::, but the selector is invalid"`;
